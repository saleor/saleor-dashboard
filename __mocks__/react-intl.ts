import { createElement, Fragment } from "react";

const useIntl = jest.fn(() => ({
  formatMessage: jest.fn(x => x.defaultMessage),
  formatDate: jest.fn(x => x),
  formatTime: jest.fn(x => x),
  formatNumber: jest.fn(x => x),
  locale: "en",
}));

const defineMessages = jest.fn(x => x);
const defineMessage = jest.fn(x => x);

const FormattedMessage = jest.fn(({ defaultMessage }: { defaultMessage: string }) =>
  createElement(Fragment, null, defaultMessage),
);

const IntlProvider = ({ children }: { children: React.ReactNode }) =>
  createElement(Fragment, null, children);

const createIntl = jest.fn(() => ({
  formatMessage: jest.fn(x => x.defaultMessage),
  formatDate: jest.fn(x => x),
  formatTime: jest.fn(x => x),
  formatNumber: jest.fn(x => x),
  locale: "en",
}));

const FormattedDate = ({ value }: { value: any }) => createElement(Fragment, null, value);

export {
  useIntl,
  defineMessages,
  defineMessage,
  FormattedMessage,
  IntlProvider,
  createIntl,
  FormattedDate,
};

// Export types for TypeScript
export type {
  IntlShape,
  MessageDescriptor,
  FormattedMessageProps,
  ReactIntlErrorCode,
} from "react-intl";
