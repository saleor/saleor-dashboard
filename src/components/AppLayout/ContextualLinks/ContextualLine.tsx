import { Box, Paragraph, TextProps } from "@saleor/macaw-ui-next";
import React from "react";

const Root = ({ children, ...rest }: TextProps) => (
  <Paragraph
    width="100%"
    color="default2"
    fontWeight="medium"
    fontSize={2}
    paddingBottom={2}
    paddingTop={1}
    {...rest}
  >
    {children}
  </Paragraph>
);

const ContextualLineLink = ({
  href,
  children,
  onClick,
}: {
  href: string;
  children: React.ReactNode;
  onClick?: () => void;
}) => (
  <Box
    as="a"
    textDecoration="underline"
    href={href}
    target="_blank"
    rel="noopener noreferrer"
    onClick={onClick}
  >
    {children}
  </Box>
);

export const ContextualLine = Object.assign(Root, {
  Link: ContextualLineLink,
});
