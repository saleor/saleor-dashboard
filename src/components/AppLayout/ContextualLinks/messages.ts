import { defineMessages } from "react-intl";

export const contextualLinks = defineMessages({
  devModePanel: {
    defaultMessage: "Learn more about {apiReference} and view {apiGuide}",
    id: "zRIsjN",
  },
  apiReference: {
    defaultMessage: "API reference",
    id: "Zvo5iu",
  },
  apiGuide: {
    defaultMessage: "API guide",
    id: "BA4leV",
  },
  webhooks: {
    defaultMessage: "Learn more about {extendingSaleor}",
    id: "zT1CvH",
  },
  extendingSaleor: {
    defaultMessage: "extending Saleor with Webhooks",
    id: "B8PvdI",
  },
  staffMembers: {
    defaultMessage: "Learn more about {userPermissions}",
    id: "19o9WS",
  },
  userPermissions: {
    defaultMessage: "User permissions",
    id: "5M+hLZ",
  },
  orders: {
    defaultMessage: "Learn more about {orderManagement}",
    id: "tl1/1E",
  },
  orderManagement: {
    defaultMessage: "order management",
    id: "91vQMc",
  },
  products: {
    defaultMessage: "Learn more about {productConfigurations}",
    id: "C9cgck",
  },
  productConfigurations: {
    defaultMessage: "product configurations",
    id: "dVk241",
  },
});
