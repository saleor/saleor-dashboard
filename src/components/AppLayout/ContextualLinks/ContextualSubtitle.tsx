import {
  API_GUIDE_DOCS,
  API_REFERENCE_DOCS,
  EXTENDING_WITH_WEBHOOKS_DOCS_URL,
  ORDER_MANAGEMENT_DOCS_URL,
  PRODUCT_CONFIGURATION_DOCS_URL,
  USER_PERMISSIONS_DOCS_URL,
} from "@dashboard/links";
import * as React from "react";
import { IntlShape } from "react-intl";

import { ContextualLine } from "./ContextualLine";
import { contextualLinks } from "./messages";

type SubtitleType =
  | "extending_saleor"
  | "product_list"
  | "order_list"
  | "dev_panel"
  | "webhooks_events"
  | "staff_members";

export const getContextualSubtitle = (type: SubtitleType, intl: IntlShape) => {
  switch (type) {
    case "staff_members":
      return intl.formatMessage(contextualLinks.staffMembers, {
        userPermissions: (
          <ContextualLine.Link href={USER_PERMISSIONS_DOCS_URL}>
            {intl.formatMessage(contextualLinks.userPermissions)}
          </ContextualLine.Link>
        ),
      });
    case "extending_saleor":
      return intl.formatMessage(contextualLinks.webhooks, {
        extendingSaleor: (
          <ContextualLine.Link href={EXTENDING_WITH_WEBHOOKS_DOCS_URL}>
            {intl.formatMessage(contextualLinks.extendingSaleor)}
          </ContextualLine.Link>
        ),
      });
    case "dev_panel":
      return intl.formatMessage(contextualLinks.devModePanel, {
        apiReference: (
          <ContextualLine.Link href={API_REFERENCE_DOCS}>
            {intl.formatMessage(contextualLinks.apiReference)}
          </ContextualLine.Link>
        ),
        apiGuide: (
          <ContextualLine.Link href={API_GUIDE_DOCS}>
            {intl.formatMessage(contextualLinks.apiGuide)}
          </ContextualLine.Link>
        ),
      });
    case "order_list":
      return intl.formatMessage(contextualLinks.orders, {
        orderManagement: (
          <ContextualLine.Link href={ORDER_MANAGEMENT_DOCS_URL}>
            {intl.formatMessage(contextualLinks.orderManagement)}
          </ContextualLine.Link>
        ),
      });
    case "product_list":
      return intl.formatMessage(
        {
          defaultMessage: "Learn more about {configurationLink}",
          id: "/noPjR",
        },
        {
          configurationLink: (
            <ContextualLine.Link href={PRODUCT_CONFIGURATION_DOCS_URL}>
              {intl.formatMessage({
                defaultMessage: "product configurations",
                id: "dVk241",
              })}
            </ContextualLine.Link>
          ),
        },
      );
    default:
      return null;
  }
};
