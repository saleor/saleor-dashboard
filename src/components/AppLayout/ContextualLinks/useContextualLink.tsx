import { useAnalytics } from "@dashboard/components/ProductAnalytics/useAnalytics";
import {
  API_GUIDE_DOCS,
  API_REFERENCE_DOCS,
  EXTENDING_WITH_WEBHOOKS_DOCS_URL,
  ORDER_MANAGEMENT_DOCS_URL,
  PRODUCT_CONFIGURATION_DOCS_URL,
  USER_PERMISSIONS_DOCS_URL,
} from "@dashboard/links";
import * as React from "react";
import { useIntl } from "react-intl";

import { ContextualLine } from "./ContextualLine";
import { contextualLinks } from "./messages";

type SubtitleType =
  | "extending_saleor"
  | "product_list"
  | "order_list"
  | "dev_panel"
  | "webhooks_events"
  | "staff_members";

const EVENT = "contextual_link_clicked";

export const useContextualLink = (type: SubtitleType) => {
  const { trackEvent: track } = useAnalytics();
  const intl = useIntl();
  const trackEvent = (type: string) => track(EVENT, { type });

  switch (type) {
    case "staff_members":
      return intl.formatMessage(contextualLinks.staffMembers, {
        userPermissions: (
          <ContextualLine.Link
            href={USER_PERMISSIONS_DOCS_URL}
            onClick={() => trackEvent("user_permissions_docs")}
          >
            {intl.formatMessage(contextualLinks.userPermissions)}
          </ContextualLine.Link>
        ),
      });
    case "extending_saleor":
      return intl.formatMessage(contextualLinks.webhooks, {
        extendingSaleor: (
          <ContextualLine.Link
            href={EXTENDING_WITH_WEBHOOKS_DOCS_URL}
            onClick={() => trackEvent("extending_saleor_docs")}
          >
            {intl.formatMessage(contextualLinks.extendingSaleor)}
          </ContextualLine.Link>
        ),
      });
    case "dev_panel":
      return intl.formatMessage(contextualLinks.devModePanel, {
        apiReference: (
          <ContextualLine.Link
            href={API_REFERENCE_DOCS}
            onClick={() => trackEvent("api_reference_docs")}
          >
            {intl.formatMessage(contextualLinks.apiReference)}
          </ContextualLine.Link>
        ),
        apiGuide: (
          <ContextualLine.Link href={API_GUIDE_DOCS} onClick={() => trackEvent("api_guide_docs")}>
            {intl.formatMessage(contextualLinks.apiGuide)}
          </ContextualLine.Link>
        ),
      });
    case "order_list":
      return intl.formatMessage(contextualLinks.orders, {
        orderManagement: (
          <ContextualLine.Link
            href={ORDER_MANAGEMENT_DOCS_URL}
            onClick={() => trackEvent("order_management_docs")}
          >
            {intl.formatMessage(contextualLinks.orderManagement)}
          </ContextualLine.Link>
        ),
      });
    case "product_list":
      return intl.formatMessage(contextualLinks.products, {
        productConfigurations: (
          <ContextualLine.Link
            href={PRODUCT_CONFIGURATION_DOCS_URL}
            onClick={() => trackEvent("product_configuration_docs")}
          >
            {intl.formatMessage(contextualLinks.productConfigurations)}
          </ContextualLine.Link>
        ),
      });
    default:
      return null;
  }
};
