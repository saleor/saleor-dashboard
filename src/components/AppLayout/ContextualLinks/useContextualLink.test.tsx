import { ContextualLine } from "@dashboard/components/AppLayout/ContextualLinks/ContextualLine";
import { renderHook } from "@testing-library/react-hooks";
import React, { ReactNode } from "react";
import { IntlProvider } from "react-intl";

import { useContextualLink } from "./useContextualLink";

jest.mock("@dashboard/components/ProductAnalytics/useAnalytics", () => ({
  useAnalytics: jest.fn().mockReturnValue({ trackEvent: jest.fn() }),
}));

const wrapper = ({ children }: { children: ReactNode }) => (
  <IntlProvider locale="EN">{children}</IntlProvider>
);

describe("useContextualLink", () => {
  it("should return correct link for staff_members", () => {
    // Act
    const { result } = renderHook(() => useContextualLink("staff_members"), { wrapper });

    // Assert
    expect(result.current).toEqual([
      "Learn more about ",
      expect.objectContaining({
        type: ContextualLine.Link,
        props: expect.objectContaining({
          href: "https://docs.saleor.io/developer/permissions#user-permissions",
          children: "User permissions",
        }),
      }),
    ]);
  });

  it("should return correct link for extending_saleor", () => {
    // Act
    const { result } = renderHook(() => useContextualLink("extending_saleor"), { wrapper });

    // Assert
    expect(result.current).toEqual([
      "Learn more about ",
      expect.objectContaining({
        type: ContextualLine.Link,
        props: expect.objectContaining({
          href: "https://docs.saleor.io/developer/extending/webhooks/overview",
          children: "extending Saleor with Webhooks",
        }),
      }),
    ]);
  });

  it("should return correct link for dev_panel", () => {
    // Act
    const { result } = renderHook(() => useContextualLink("dev_panel"), { wrapper });

    // Assert
    expect(result.current).toEqual([
      "Learn more about ",
      expect.objectContaining({
        type: ContextualLine.Link,
        props: expect.objectContaining({
          href: "https://docs.saleor.io/api-reference/",
          children: "API reference",
        }),
      }),
      " and view ",
      expect.objectContaining({
        type: ContextualLine.Link,
        props: expect.objectContaining({
          href: "https://docs.saleor.io/api-usage/overview",
          children: "API guide",
        }),
      }),
    ]);
  });

  it("should return correct link for order_list", () => {
    // Act
    const { result } = renderHook(() => useContextualLink("order_list"), { wrapper });

    // Assert
    expect(result.current).toEqual([
      "Learn more about ",
      expect.objectContaining({
        type: ContextualLine.Link,
        props: expect.objectContaining({
          href: "https://docs.saleor.io/developer/checkout/order-status",
          children: "order management",
        }),
      }),
    ]);
  });

  it("should return correct link for product_list", () => {
    // Act
    const { result } = renderHook(() => useContextualLink("product_list"), { wrapper });

    // Assert
    expect(result.current).toEqual([
      "Learn more about ",
      expect.objectContaining({
        type: ContextualLine.Link,
        props: expect.objectContaining({
          href: "https://docs.saleor.io/developer/products/configuration",
          children: "product configurations",
        }),
      }),
    ]);
  });

  it("should return correct link for gift_cards", () => {
    // Act
    const { result } = renderHook(() => useContextualLink("gift_cards"), { wrapper });

    // Assert
    expect(result.current).toEqual([
      "Learn more about ",
      expect.objectContaining({
        type: ContextualLine.Link,
        props: expect.objectContaining({
          href: "https://docs.saleor.io/developer/gift-cards",
          children: "gift cards",
        }),
      }),
    ]);
  });
});
