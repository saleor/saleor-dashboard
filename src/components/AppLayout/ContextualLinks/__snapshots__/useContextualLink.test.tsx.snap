// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useContextualLink should return correct link for dev_panel 1`] = `
Object {
  "all": Array [
    Array [
      "Learn more about ",
      <ContextualLineLink
        href="https://docs.saleor.io/api-reference/"
        onClick={[Function]}
      >
        API reference
      </ContextualLineLink>,
      " and view ",
      <ContextualLineLink
        href="https://docs.saleor.io/api-usage/overview"
        onClick={[Function]}
      >
        API guide
      </ContextualLineLink>,
    ],
  ],
  "current": Array [
    "Learn more about ",
    <ContextualLineLink
      href="https://docs.saleor.io/api-reference/"
      onClick={[Function]}
    >
      API reference
    </ContextualLineLink>,
    " and view ",
    <ContextualLineLink
      href="https://docs.saleor.io/api-usage/overview"
      onClick={[Function]}
    >
      API guide
    </ContextualLineLink>,
  ],
  "error": undefined,
}
`;

exports[`useContextualLink should return correct link for extending_saleor 1`] = `
Object {
  "all": Array [
    Array [
      "Learn more about ",
      <ContextualLineLink
        href="https://docs.saleor.io/developer/extending/webhooks/overview"
        onClick={[Function]}
      >
        extending Saleor with Webhooks
      </ContextualLineLink>,
    ],
  ],
  "current": Array [
    "Learn more about ",
    <ContextualLineLink
      href="https://docs.saleor.io/developer/extending/webhooks/overview"
      onClick={[Function]}
    >
      extending Saleor with Webhooks
    </ContextualLineLink>,
  ],
  "error": undefined,
}
`;

exports[`useContextualLink should return correct link for gift_cards 1`] = `
Object {
  "all": Array [
    Array [
      "Learn more about ",
      <ContextualLineLink
        href="https://docs.saleor.io/developer/gift-cards"
        onClick={[Function]}
      >
        gift cards
      </ContextualLineLink>,
    ],
  ],
  "current": Array [
    "Learn more about ",
    <ContextualLineLink
      href="https://docs.saleor.io/developer/gift-cards"
      onClick={[Function]}
    >
      gift cards
    </ContextualLineLink>,
  ],
  "error": undefined,
}
`;

exports[`useContextualLink should return correct link for order_list 1`] = `
Object {
  "all": Array [
    Array [
      "Learn more about ",
      <ContextualLineLink
        href="https://docs.saleor.io/developer/checkout/order-status"
        onClick={[Function]}
      >
        order management
      </ContextualLineLink>,
    ],
  ],
  "current": Array [
    "Learn more about ",
    <ContextualLineLink
      href="https://docs.saleor.io/developer/checkout/order-status"
      onClick={[Function]}
    >
      order management
    </ContextualLineLink>,
  ],
  "error": undefined,
}
`;

exports[`useContextualLink should return correct link for product_list 1`] = `
Object {
  "all": Array [
    Array [
      "Learn more about ",
      <ContextualLineLink
        href="https://docs.saleor.io/developer/products/configuration"
        onClick={[Function]}
      >
        product configurations
      </ContextualLineLink>,
    ],
  ],
  "current": Array [
    "Learn more about ",
    <ContextualLineLink
      href="https://docs.saleor.io/developer/products/configuration"
      onClick={[Function]}
    >
      product configurations
    </ContextualLineLink>,
  ],
  "error": undefined,
}
`;

exports[`useContextualLink should return correct link for staff_members 1`] = `
Object {
  "all": Array [
    Array [
      "Learn more about ",
      <ContextualLineLink
        href="https://docs.saleor.io/developer/permissions#user-permissions"
        onClick={[Function]}
      >
        User permissions
      </ContextualLineLink>,
    ],
  ],
  "current": Array [
    "Learn more about ",
    <ContextualLineLink
      href="https://docs.saleor.io/developer/permissions#user-permissions"
      onClick={[Function]}
    >
      User permissions
    </ContextualLineLink>,
  ],
  "error": undefined,
}
`;

exports[`useContextualLink should return correct link for webhooks_events 1`] = `
Object {
  "all": Array [
    null,
  ],
  "current": null,
  "error": undefined,
}
`;
